File: FingoDriver/src/main/java/com/yelloco/fingodriver/FingoSDK.java
package com.yelloco.fingodriver;

import android.annotation.SuppressLint;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbManager;
import android.util.Log;

import com.yelloco.fingodriver.callbacks.FingoRequestLogger;
import com.yelloco.fingodriver.enums.FingoErrorCode;
import com.yelloco.fingodriver.enums.StorageKey;
import com.yelloco.fingodriver.utils.FingoParams;
import com.yelloco.fingodriver.utils.FingoUsbManager;
import com.yelloco.fingodriver.utils.Storage;
import com.yelloco.fingodriver.utils.UsbReceiver;

import org.greenrobot.eventbus.EventBus;

import java.util.HashMap;

public class FingoSDK
{
    // Constants
    private static final String TAG = FingoSDK.class.getSimpleName();

    // Memebers
    protected static boolean sdkInitialized;
    @SuppressLint("StaticFieldLeak")
    private static FingoUsbManager fingoUsbManager;
    public static FingoRequestLogger fingoRequestLogger;

    private FingoSDK(){
    }

    public static FingoErrorCode initialize(Context context){
        if(sdkInitialized){
            Log.w(TAG, "FingoSDK Already INITIALIZED");
            return FingoErrorCode.H1_DRIVER_INITIALIZED;
        }
        if(context == null){
            Log.e(TAG, "FingoSDK received NULL context");
            return FingoErrorCode.H1_UNEXPECTED;
        }

        Storage.getInstance().initialize(context);

        fingoUsbManager = new FingoUsbManager(context);
        FingoErrorCode usbInitStatus = fingoUsbManager.initialize();

        if(usbInitStatus != FingoErrorCode.H1_OK){
            return usbInitStatus;
        }

        sdkInitialized = true;
        Log.d(TAG, "Fingo SDK Initialized");

        EventBus.getDefault().register(FingoPayDriver.getInstance());

        fingoUsbManager.checkAttachedDevices();

        return FingoErrorCode.H1_OK;
    }

    public static FingoErrorCode initialize(Context context, FingoRequestLogger fingoRequestLogger){
        FingoSDK.fingoRequestLogger = fingoRequestLogger;
        return FingoSDK.initialize(context);
    }

    public static FingoErrorCode setFingoParams(FingoParams fingoParams){
        FingoErrorCode fingoParamsErrorCode = fingoParams.validate();

        Log.i(TAG, "setFingoParams validation result: " + fingoParamsErrorCode);
        return fingoParamsErrorCode;
    }

    public static FingoParams getFingoParams(){
        return FingoParams.getStoredParams();
    }

    public static String about(){
        return FingoSDK.about(null);
    }

    public static String about(String uniqueID){
        if(FingoConstants.KAN_UNIQUE_ID.equals(uniqueID)){
            return "The Library And It's Rights Belongs to KAN (www.kan4u.com)\nplease contact info@kan4u.com for inquiries";
        }
        else {
            return "Disclaimer and Its rights are returned here";
        }
    }

    public static void destroy() {
        Log.d(TAG, "Is Fingo Initialized: " + sdkInitialized);
        if(! sdkInitialized)
            return;

        sdkInitialized = false;
        EventBus.getDefault().unregister(FingoPayDriver.getInstance());

        fingoUsbManager.destroy();
    }

    public static boolean isSdkInitialized() {
        return sdkInitialized;
    }
}

File: FingoDriver/src/main/java/com/yelloco/fingodriver/FingoSDK.kt
package com.yelloco.fingodriver

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log
import com.yelloco.fingodriver.callbacks.FingoRequestLogger
import com.yelloco.fingodriver.enums.FingoErrorCode
import com.yelloco.fingodriver.enums.StorageKey
import com.yelloco.fingodriver.utils.FingoParams
import com.yelloco.fingodriver.utils.FingoUsbManager
import com.yelloco.fingodriver.utils.Storage
import com.yelloco.fingodriver.utils.Storage.storeBoolean
import org.greenrobot.eventbus.EventBus

object FingoSDK
{
    // Constants
    private val TAG = FingoSDK::class.java.simpleName

    // Memebers
    var isSdkInitialized = false
        internal set

    @SuppressLint("StaticFieldLeak")
    private var fingoUsbManager: FingoUsbManager? = null

    var fingoRequestLogger: FingoRequestLogger? = null

    fun initialize(context: Context?): FingoErrorCode {
        if (isSdkInitialized) {
            Log.w(TAG, "FingoSDK Already INITIALIZED")
            return FingoErrorCode.H1_DRIVER_INITIALIZED
        }
        if (context == null) {
            Log.e(TAG, "FingoSDK received NULL context")
            return FingoErrorCode.H1_UNEXPECTED
        }

        Storage.initialize(context)
        storeBoolean(StorageKey.PARAMS_STATUS.name, false)
        fingoUsbManager = FingoUsbManager(context)
        val usbInitStatus = fingoUsbManager!!.initialize()
        if (usbInitStatus !== FingoErrorCode.H1_OK) {
            return usbInitStatus
        }
        isSdkInitialized = true
        Log.d(TAG, "Fingo SDK Initialized")
        EventBus.getDefault().register(FingoPayDriver)
        fingoUsbManager!!.checkAttachedDevices()
        return FingoErrorCode.H1_OK
    }

    fun initialize(context: Context?, fingoRequestLogger: FingoRequestLogger?): FingoErrorCode {
        FingoSDK.fingoRequestLogger = fingoRequestLogger
        return initialize(context)
    }

    fun setFingoParams(fingoParams: FingoParams): FingoErrorCode {
        val fingoParamsErrorCode = fingoParams.validate()
        if (fingoParamsErrorCode === FingoErrorCode.H1_OK) {
            fingoParams.loadParams()
        } else {
            storeBoolean(StorageKey.PARAMS_STATUS.name, false)
        }
        Log.i(TAG, "setFingoParams validation result: $fingoParamsErrorCode")
        return fingoParamsErrorCode
    }

    fun about(): String {
        return about(null)
    }

    fun about(uniqueID: String?): String {
        return if (FingoFactory.Constants.KAN_UNIQUE_ID == uniqueID) {
            "The Library And It's Rights Belongs to KAN (www.kan4u.com)\nplease contact info@kan4u.com for inquiries"
        } else {
            "Disclaimer and Its rights are returned here"
        }
    }

    fun destroy() {
        Log.d(TAG, "Is Fingo Initialized: $isSdkInitialized")
        if (!isSdkInitialized) return
        isSdkInitialized = false
        EventBus.getDefault().unregister(FingoPayDriver)
        fingoUsbManager!!.destroy()
    }
}